File: C++/109. Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        vector<int> res;
        while(head!=NULL){
            res.push_back(head->val);
            head=head->next;
        }
        return sol(0,res.size()-1,res);
    }
    TreeNode* sol(int l, int r, vector<int> &res){
        if(l>r){
            return 0;
        }
        int m = (l+r)/2;
        TreeNode* root= new TreeNode(res[m]);
        root->left=sol(l,m-1,res);
        root->right=sol(m+1,r,res);
        return root;
    }
};

File: C#/109. Convert Sorted List to Binary Search Tree.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SortedListToBST(ListNode head) {
        List<int> res = new List<int>();
        while(head!=null){
            res.Add(head.val);
            head=head.next;
        }
        return sol(0,res.Count-1,res);
    }
    public TreeNode sol(int l, int r, List<int> res){
        if(l>r){
            return null;
        }
        int m = (l+r)/2;
        TreeNode root= new TreeNode(res[m]);
        root.left=sol(l,m-1,res);
        root.right=sol(m+1,r,res);
        return root;
    }
}
